src/java.base/share/classes/java/lang/reflect/code/type/JavaType.java

```diff
@@ -172,26 +154,28 @@ private static JavaType ofNominalDescriptorStringInternal(String descriptor, int
                 case 'F' -> JavaType.FLOAT;
                 case 'D' -> JavaType.DOUBLE;
                 case 'Z' -> JavaType.BOOLEAN;
-                case 'L' -> {
-                    // La.b.c.Class;
-                    String typeName = descriptor.substring(i + 1, descriptor.length() - 1).replace('/', '.');
-                    yield new ClassType(typeName);
-                }
                 default -> throw new InternalError();
             };
+        } else if (d.isArray()) {
+            return array(type(d.componentType()));
+        } else {
+            // class
+            return new ClassType(d, List.of());
         }
     }
 
-    static JavaType type(JavaType t, JavaType... typeArguments) {
-        return type(t, List.of(typeArguments));
+    static ClassType parameterized(JavaType type, JavaType... typeArguments) {
```

Comment:
Ideally this would take a `ClassType` as base type (or, even a `ClassDesc`/`Class<?>`). That would require some more changes at the use site.